### ISSUE TITLE
actions.payment.execute completed but the payment was not made

### ISSUE THREAD

**Issue by mpserellis** [paypal/paypal-checkout#402]
Firstly, I am afraid I cannot reproduce this error, however, it happened 3 times in production so I am hoping you may be able to help with extra steps my code should be taking to handle this. I must also say that there have been many successful transactions before and after.

The customer clicked the PayPal button and completed payment using guest checkout. onAuthorize was fired which, as per the documentation, calls the execute function and waits for success.

`              onAuthorize: function (data, actions) {
                return actions.payment.execute().then(function () {
                  $watchForm.empty().append('<h3>Thank you</h3><p>You should receive confirmation of your payment from PayPal.</p>');
                });
              },
`

My client checked their PayPal account and there was no transaction for this payment. The customer tried again and finally on a different day my client tried on behalf of the customer. Each time the payment didn't appear in PayPal. That is when they contacted me and luckily they had kept the web page loaded. I checked Chrome Dev Tools and there were 2 errors from the execute URL, the first was a 401 error which happens quite often and I see you even have some code that automatically retries, but then the retry returned a 400 error.

Despite the 400 error, the "then" function was called as if the execute had been successful. I know this because of the confirmation message rendered to the page. This is the only time it is shown.

![image001](https://user-images.githubusercontent.com/8006172/28117703-6c579d34-6706-11e7-8832-d0306f0e0570.png)

Please could you let me know if there was any way my code could catch this error or any checks I can make in the "then" function to ensure payment was executed successfully.

Thank you
Mark

**Comment by mpserellis** [comment]
I have some more information regarding this specific scenario. My client sent the customer an invoice via PayPal and when they tried to pay using the same credit card it was declined. Could this have been what caused the 400 response code?

I have been trying to get negative testing working in order to try replicate the issue but to no avail. I have followed the instructions to turn on negative testing in the sandbox account and have sent the required amount totals but each time process completes successfully. Can anyone help with negative testing using checkout.js?

Thank you

**Comment by bluepnume** [comment]
> Despite the 400 error, the "then" function was called as if the execute had been successful. I know this because of the confirmation message rendered to the page. This is the only time it is shown.

Yeah this is a problem. But I'm not sure what the cause is. We're treating `>= 400` as an error case for any http calls:

https://github.com/paypal/paypal-checkout/blob/master/src/lib/http.js#L73-L85

Could you also share the response body for that /execute call? That should also give a clue.

**Comment by bluepnume** [comment]
Here's a quick test I ran:

<img width="631" alt="screen shot 2017-07-18 at 11 50 37 am" src="https://user-images.githubusercontent.com/1219075/28334238-5f3c8440-6baf-11e7-976c-bdabc6c606a2.png">

**Comment by mpserellis** [comment]
Thank you for the response. I am really sorry, I did not copy the response body, I had limited time viewing my client's PC remotely and with the situation I didn't think to do that.

JavaScript is not my first language so this may be completely irrelevant but I found the below code using dev tools:

xcomponent__ppbutton__4... > webpack:///js/button.js

line 909

`	            // Turn 400s with an ack into a 'success' so it makes it to the standard handlers
	            if (res && res.status && res.status.toString() === '400' && res.data && res.data.ack) {
	                return res;
	            }`

I will continue to try get negative testing working. Thank you.
